### ISSUE TITLE
Issue of sampling procedure

### ISSUE THREAD

**Issue by taoshen58** [RiaanZoetmulder/Master-Thesis#2]
First of all, I appreciate you can pubilc the tensorflow code of paper _Rationalizing Neural Prediction_. 

I have transfer the code to tensorflow 1.2 and python3, but I have a question about the code of sampling procedure. Specifically, in file _advanced_layers.py_, class _Z_layer_, function _sample_all_. The Code is here:
`h0 = tf.zeros((x.get_shape()[1], self._n_hidden*(self.rlayer._order+1)), dtype = tf.float32)`
`z0 = tf.zeros((x.get_shape()[1],), dtype=tf.float32)  # [256,1]`
`h, z = tf.scan( self.sample, x, initializer = [ h0, tf.expand_dims(z0, 1 )])`
The shape of `x` is [batch_size, max_len, hidden_num], and I think the `tf.scan` should be performed in the second dim of `x`, which means the `x` should be transposed  using `tf.transpose(x, [1, 0, 2])` before the code displayed above. 

I have no idea if this issue is caused by different version of tensorflow, or you mean to do this, but I'm looking forward to your reply, thanks a lot.

**Comment by RiaanZoetmulder** [comment]
Hi!
Thanks for taking the time to port the code! You must forgive me, I am a little out of date when it comes to this code, so be critical.

So we have an input tensor 'x' --> [batch_size, max_len, embedding_dim]

If we wish to sample from this we must then get a tensor z --> [batch_size, max_len]

I believe we do this by scanning over max_len, batch_size times. And then multiplying it elementwise with the input tensor x. 

I do believe that this issue can arise because theano and tensorflow switch some of the dimensions around like they do in images. But I am not sure.

**Comment by taoshen58** [comment]
As the shown in the code, `tf.scan` unpack `x` along the first dim, namely the batch, the it comes that you just process, or sample the instance one by one without temporal consideration... And then, you also use a RCNN along the batch, so it makes no sense...

**Comment by RiaanZoetmulder** [comment]
Isn't this exactly what you want?

for x in batch:
- multiply x with w1 and h with w2 such that you end up with pz_t.shape[max_len, 1]
- sample using pz_t and get z_t
- get the next hidden state by concatenating z_t and x_t , xz_t = [max_len, embed_size + 1] using the RCNN.

The temporal consideration is enforced by the loss for as far as I know.

**Comment by taoshen58** [comment]
I also found some issue in the code which is used to encode contextual information, so I think you got a misunderstanding. You can recall the definition of RNN or LSTM which performs the iteration along the sequence, namely the `max_len` in your code. Actually, all the iteration in your code should be along the sequence to ensure the conditional probability p(z_t|x, z_1, ....z_{t-1}) as should in the paper. I  will change the code and I can send it to you if you are interesting.

You know what, because the batch_size and max_len is equal in your code, this does not invoke errors when running the code.

Finally, maybe this really is caused by version difference or my misunderstanding. I hope you can forgive me for bothering you.

**Comment by RiaanZoetmulder** [comment]
You are not bothering me, I'd be very interested to know whether I made a mistake!

**Comment by taoshen58** [comment]
I just found that the arguements `x` and `y` you passed into placeholders of tensorflow have different batch number. I use:
`print(bx.shape, by.shape)`
before feed_dict passed into network, then get:
`(115, 256) (256, 5)`

**Comment by angus1095** [comment]
Hi guys, I'm having the same problem with tf.scan() in models.py when running this code.
I'd be grateful if you commit any bugfixes because I haven't been able to fix it!
Thanks

**Comment by taoshen58** [comment]
Hey, I checked the code in original repo with theano, and I found the shape of input tensor (i.e. placeholder in tensorflow) is defined as `[max_len, batch_size]` in approximate line 37, 
`# len*batch`
`x = self.x = T.imatrix()`
whereas the shape in your code is `[batch_size, max_len]`, the coincidence is that your batch size is equal to the max_len. I think this is why your code can output correct result. I hope I do help for you~ Thanks again for you tensorflow code release~
